src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -208,37 +211,58 @@ maxPacketSize = 256 * 2 ^ (20 :: Int) - 128 * 2 ^ (10 :: Int)
 -- @since 0.1.0.0
 makePacketSender ::
   MonadUnliftIO m =>
-  -- | The maximum packet payload size. This must be a 'Word32' value greater 
+  -- | The maximum packet payload size. This must be a 'Word32' value greater
   -- than or equal to 'minPacketSize' and less than 'maxPacketSize'. Otherwise,
   -- a 'PacketSizeError' exception will be thrown.
   Word32 ->
+  -- | Optional rate limit for publishing as bytes per second.
+  Maybe Word32 ->
   (ByteString -> m ()) ->
   ByteString ->
   m ()
-makePacketSender limit publish message = do
-  unless (isValidPacketSize limit) do
-    throwIO (PacketSizeError $ toInteger limit)
+makePacketSender packetSizeLimit mRateLimit publish message = do
+  unless (isValidPacketSize packetSizeLimit) do
+    throwIO (PacketSizeError $ toInteger packetSizeLimit)
 
   if fromIntegral (ByteString.length message) <= maxPayloadSize
     then do
-      -- Perform a single publish on the current thread in case that the 
+      -- Perform a single publish on the current thread in case that the
       -- @message@ length is less than the maximum payload size.
       let packet = Packet message 0 1
       publish (wireWrapPacket' packet)
-    else do
-      caps <- liftIO getNumCapabilities
-      jobs <- newTVarIO 0
-
-      replicateConcurrently_ caps $ fix \next ->
-        atomically (takeJobId jobs) >>= \case
-          Nothing -> pure ()
-          Just jobid -> do
-            let packet = makePacket jobid
-            publish (wireWrapPacket' packet)
-            next
+    else case mRateLimit of
+      Nothing -> do
+        caps <- liftIO getNumCapabilities
+        jobs <- newTVarIO 0
+
+        replicateConcurrently_ caps $ fix \next ->
+          atomically (takeJobId jobs) >>= \case
+            Nothing -> pure ()
+            Just jobid -> do
+              let packet = makePacket jobid
+              publish (wireWrapPacket' packet)
+              next
+      Just rateLimit -> do
+        -- Throughput is rate limited on a 1 second interval and we don't subdivide packets to comply
+        -- with the rate limit, so the payload size must be less than or equal to the rate limit.
+        -- NB: Since `maxPayloadSize` is already at least 1, this ensures that `rateLimit` is not 0.
+        unless (maxPayloadSize <= rateLimit) $
+          throwIO (PacketSizeError $ toInteger packetSizeLimit)
+
+        -- Convert `maxPayloadSize` and `rateLimit` to Integer first to avoid overflow. Also,
+        -- the above condition that `maxPayloadSize` <= `rateLimit` ensures that the period is
+        -- between 0 and 1,000,000 microseconds, so the final conversion Integer -> Int is always safe.
+        let packetSendPeriod :: Int
+            packetSendPeriod = fromIntegral @Integer @Int $ fromIntegral maxPayloadSize * 1_000_000 `div` fromIntegral rateLimit
+
+        forM_ [0 .. njobs] $ \jobid -> do
+          timer <- async $ threadDelay packetSendPeriod
```

Comment:
Could this be accomplished with [`TSem`](https://hackage.haskell.org/package/stm-2.5.1.0/docs/Control-Concurrent-STM-TSem.html#t:TSem)? You could have a separate thread running thats loop `packetSendPeriod` that is bound by the total number of bytes to be sent. This loop would just accumulate the number of bytes that can be published into the [`TSem`](https://hackage.haskell.org/package/stm-2.5.1.0/docs/Control-Concurrent-STM-TSem.html#t:TSem). Then, in the loop running on a separate thread handling the packet publishes here

```haskell
        replicateConcurrently_ caps $ fix \next ->
          atomically (takeJobId jobs) >>= \case
            Nothing -> pure ()
            Just jobid -> do
              let packet = makePacket jobid
              publish (wireWrapPacket' packet)
              next
```

and a function which can take `n` bytes bounded by `maxPayloadSize`, roughly something like:

```haskell
waitTSemN :: TSem () -> Integer -> STM ()
waitTSemN (TSem sem) limit = do 
  i <- readTVar sem 
  when (i <= 0) retry
  let ret = min limit i
  writeTVar sem (i - ret)
  pure ret
```

and incorporate into the packet publish with something like `makePacket jobid =<< atomically takeTSemN`. The loop accumulating into the rate-limiting `TSem` and the packet publishing loop could be run in parallel with `concurrently_`.

I think the big advantage to handling throttling this way is that the logic for the throttling and the logic for sending packets is kept disjoint, removing the additional case we're handling here with `Maybe`.

Reply:
I like the idea of separating the throttling and sending logic, but I'm not quite following your suggestion here using `TSem`. Semaphores are really just counters. You can't read/write data to them.

Reply:
Yeah I was just providing a rough idea, maybe `TSem` specifically doesn't make sense in this case. The core idea I was trying to convey was just to have a `TVar Int` that carries the number of bytes that can be sent before the packet sender gets throttled, i.e. if `0 <- readTVar (var :: TVar Int)` then the packet sender blocks until `readTVar var > 0`. A separate thread could add some number of bytes to `var :: TVar Int` every `packetSendPeriod` to unblock the publishing thread until all bytes are sent.